### PR TITLE
refactor add payment dynamic

### PR DESCRIPTION
**Pull request by phivh**
### Details
- Added one window listener that observed resize event and re-set position of the menu, there is a reason that needs to migrate the listener into dimension method.

### Fixed Issues

$ [#7385](https://github.com/Expensify/App/issues/7385)

### Tests

- [x] Tested by access to the payment page from the setting page.

### QA Steps

- [x] Login on web / desktop app
- [x] Select profile > payments
- [x] Select add payment method
- [x] Change screen width
- [x] Verify

### Tested On

- [x] Web
- [ ] Mobile Web
- [x] Desktop
- [ ] iOS
- [ ] Android

#### Web

https://user-images.githubusercontent.com/5449119/154851435-85db3ef8-618a-40d2-bb15-f1bd8fbef35b.mp4

#### Mobile Web

NA
#### Desktop

https://user-images.githubusercontent.com/5449119/152561282-dffec0b4-24e0-4a76-b9a2-4b98856d6065.mp4

#### iOS

NA
#### Android

NA

### CODE REVIEW

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
@phivh small issue here - payment popover jumps down on resizing.

https://user-images.githubusercontent.com/29673073/155041131-62b56ebc-22a9-409f-9d5f-f921e2d21819.mp4

**Comment by phivh** [comment]
@rushatgabhane @deetergp updated!

**Comment by rushatgabhane** [comment]
@phivh could you please post a quick explaination on what fixed the resizing bug? Thanks

File: src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js [hunk 1]
```diff
@@ -52,10 +53,32 @@ class BasePaymentsPage extends React.Component {
         this.deletePaymentMethod = this.deletePaymentMethod.bind(this);
         this.hidePasswordPrompt = this.hidePasswordPrompt.bind(this);
         this.navigateToTransferBalancePage = this.navigateToTransferBalancePage.bind(this);
+        this.setMenuPosition = this.setMenuPosition.bind(this);
+        this.dimensionsSubscription = null;
```
Comment (rushatgabhane) [inline src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js:57]: This isn't needed, and can be removed

File: src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js [hunk 2]
```diff
@@ -52,10 +53,32 @@ class BasePaymentsPage extends React.Component {
...
     componentDidMount() {
         PaymentMethods.getPaymentMethods();
+        if (this.props.shouldListenForResize) {
+            this.dimensionsSubscription = Dimensions.addEventListener('change', this.setMenuPosition);
+        }
+    }
+
+    componentWillUnmount() {
+        this.setState({
+            addPaymentMethodButton: null,
```
Comment (rushatgabhane) [inline src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js:69]: Are you sure we need to put this in state? wouldn't a class variable suffice?

File: src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js [hunk 3]
```diff
@@ -52,10 +53,32 @@ class BasePaymentsPage extends React.Component {
...
+        if (this.props.shouldListenForResize && this.dimensionsSubscription) {
+            this.dimensionsSubscription.remove();
+        }
+    }
+
+    setMenuPosition() {
+        if (!this.state.addPaymentMethodButton) {
+            return;
+        }
+        const btnPosition = getClickedElementLocation(this.state.addPaymentMethodButton);
```
Comment (rushatgabhane) [inline src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js:80]: rename to `buttonPosition`

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
@phivh really minor changes

**Review by deetergp — CHANGES_REQUESTED** [review]
I don't have anything to add to what @rushatgabhane has requested already. Let's get those changes made and this will be good to go! 👍

File: src/components/KYCWall/BaseKYCWall.js [hunk 4]
```diff
@@ -20,26 +20,41 @@ class KYCWall extends React.Component {
...
 
         this.continue = this.continue.bind(this);
+        this.setMenuPosition = this.setMenuPosition.bind(this);
+        this.dimensionsSubscription = null;
 
         this.state = {
             shouldShowAddPaymentMenu: false,
             anchorPositionTop: 0,
             anchorPositionLeft: 0,
+            transferBalanceButton: null,
```
[the remark this answers is not in the document]
Comment (phivh) [reply src/components/KYCWall/BaseKYCWall.js:30]: The position of the menu belongs to position of the button that user clicking on. Thus, when use listener we need to re-assign this button when any click event happens, the possible way to re-assign any variables to the listener in this case needs to use state.

File: src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js [hunk 5]
```diff
@@ -42,6 +42,7 @@ class BasePaymentsPage extends React.Component {
             formattedSelectedPaymentMethod: {},
             anchorPositionTop: 0,
             anchorPositionLeft: 0,
+            addPaymentMethodButton: null,
```
[the remark this answers is not in the document]
Comment (phivh) [reply src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js:45]: I've explained the reason for the comment below.

**Comment by phivh** [comment]
@rushatgabhane @deetergp updated.
There are not many changes from the previous PR, the only things that only update window listener to Dimension of RN.

**Review by rushatgabhane — APPROVED** [review]
@deetergp LGTM! 🎉️

**Review by deetergp — APPROVED** [review]
LGTM

File: src/components/KYCWall/BaseKYCWall.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply src/components/KYCWall/BaseKYCWall.js:29]: I know this is old, but I don't understand this explanation and it seems wrong. @rushatgabhane please try to address all comments before approving 🙇‍♂️

[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/components/KYCWall/BaseKYCWall.js:29]: @marcaaron im sorry about it
> please try to address all comments before approving

yess!

**Pull request merged**